### PR TITLE
fix Issue 15399 - unaligned pointers are not @safe

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Comment by yebblies** [comment]
Taking the address too?

**Comment by WalterBright** [comment]
Toto too!

File: test/fail_compilation/test15399.d [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    p = s2.ptr;
+    s1.ptr = null;
+    s2.ptr = null;
+    int** pp = &s1.ptr;
+    pp = &s2.ptr;
+    bar(s1.ptr);
+    bar(s2.ptr);
+    sinister(s1.ptr);
+    sinister(s2.ptr);
+}
```
Comment (ZombineDev) [inline test/fail_compilation/test15399.d]: What about `int x = *s1.ptr` and `*s1.ptr = 42`?

Comment (WalterBright) [inline test/fail_compilation/test15399.d]: It's writing a new value to the misaligned pointer that is the problem, not dereferencing it.

Comment (ZombineDev) [inline test/fail_compilation/test15399.d]: Alright, tnx for the explanation.

File: src/expression.d [hunk 2]
```diff
@@ -80,6 +80,43 @@ void emplaceExp(T : UnionExp)(T* p, Expression e)
...
+ *      sc = scope
+ *      e = expression to check
+ *      msg = error message string
+ * Returns:
+ *      true if error
+ */
+
+bool checkMisalignedPointer(Scope* sc, Expression e, string msg)
+{
+    if (e.op == TOKdotvar)
```
Comment (andralex) [inline src/expression.d]: This is a prime example of that pattern that should be replaced with:

```D
if (e.op != TOKdotvar)
    return false;
... interesting branch ...
```

Clarifies the charter of the scope and unindents everything that's interesting.

**Comment by andralex** [comment]
ok modulo pet peeve

**Comment by WalterBright** [comment]
@andralex rebased, working now

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**